### PR TITLE
SUIT: Introduction of a payload storage API for SUIT manifest payloads

### PR DESCRIPTION
**Pull request by bergzand**
### Contribution description

This PR adds an API for generic SUIT payload handling together with two implementations.

The storage driver API allows for creating multiple backends, each possibly servicing
multiple locations. An example of this is a VFS storage backend. This backend
could service multiple file locations on a filesystem.

Two implementations are included in this PR:
 - A RAM based backend. Payloads are stored in a series of ram buffers. Main use of this is testing, but I think it could also be used to store payloads in the backup memory of some microcontrollers.
 - riotboot flashwrite: A backend based on the riotboot flashwrite module. This backend replaces the current direct calls to the riotboot module.

Planned:
 - VFS based: A VFS-based backend to write files on the filesystem via SUIT updates. This could be used to update configuration files via SUIT.
 - MTD: Direct MTD writes can be used for a two-stage update process where the update is first written to an external flash and moved to the internal flash by the bootloader.

### Testing procedure

Both `tests/suit_manifest` and `examples/suit_update` must still behave as usual.
Please also give the API docs in `sys/include/suit/storage.h` a read to check if it is somewhat clear to an uninformed reader.

### Issues/PRs references

A pair of patches riotboot are included, I can split them out if preferred.

Depends on #15092

### CODE REVIEW

**Comment by bergzand** [comment]
The additional abstractions included here cause the `examples/suit_update` firmware to increase by 832 bytes. (`LOG_LEVEL=LOG_NONE`, `RIOT_CI_BUILD=1`)

File: sys/include/suit/transport/mock.h [hunk 1]
```diff
@@ -0,0 +1,61 @@
...
+extern "C" {
+#endif
+
+/**
+ * @brief Mock payload.
+ */
+typedef struct {
+    const uint8_t *buf; /**< Ptr to the memory space containing the payload */
+    size_t len;         /**< Length of the payload in bytes */
+}suit_transport_mock_payload_t;
```
Comment (kaspar030) [inline sys/include/suit/transport/mock.h:41]:
```suggestion
} suit_transport_mock_payload_t;
```

**Comment by kaspar030** [comment]
Tests are failing, looks like due to changed output.

**Comment by bergzand** [comment]
I've restored the essential and useful bits of output

**Comment by bergzand** [comment]
Test succeeds here on the nrf52840dk

**Comment by kaspar030** [comment]
- please squash!

**Review by kaspar030 — APPROVED** [review]
ACK.

Code looks good. Nicely documented.
The increased code size is a bit painful, but we discussed offline that some of it can be mitigated.

The added potential regarding upgrade flexibility is huge.

**Comment by bergzand** [comment]
Squashed!

File: sys/include/suit.h [hunk 2]
```diff
@@ -189,12 +199,20 @@ typedef struct {
 #define SUIT_COMPONENT_STATE_FINALIZED     (1 << 3) /**< Component successfully installed */
 /** @} */
 
+/**
+ * @brief   Forward declaration for storage struct
+ *
+ * Breaks a dependency chain
+ */
+typedef struct suit_storage suit_storage_ref_t;
```
Comment (bergzand) [inline sys/include/suit.h:207]: I've renamed this typedef to silence the LLVM error.

**Comment by benpicco** [comment]
Only the usual tests are failing

 - `run_test/tests/mtd_mapper/esp32-wroom-32:gnu`
 - `run_test/tests/pkg_littlefs2/esp32-wroom-32:gnu`
 - `run_test/tests/lwip_sock_tcp/esp32-wroom-32:gnu`
 - `run_test/tests/pkg_spiffs/esp32-wroom-32:gnu`
 - `run_test/tests/pkg_libhydrogen/samr21-xpro:llvm`
 - `run_test/tests/pkg_littlefs/esp32-wroom-32:gnu`

**Comment by kaspar030** [comment]
Let's go.

**Comment by bergzand** [comment]
Thanks!

**Comment by fjmolinas** [comment]
`examples/suit_update` is not passing for me on `iotlab-m3`, it fails with the following error, any ideas?

```
make[1]: Leaving directory '/home/francisco/workspace/RIOT3/examples/suit_update'
suit: received URL: "coap://[fd00:dead:beef::1]/fw/iotlab-m3/suit_update-riot.suit_signed.1603730758.bin"
suit_coap: trigger received
suit_coap: downloading "coap://[fd00:dead:beef::1]/fw/iotlab-m3/suit_update-riot.suit_signed.1603730758.bin"
suit_coap: got manifest with size 503
suit: verifying manifest signature
suit: validated manifest version
)riotboot_hdr_validate: riotboot_hdr magic number invalid
Manifest seq_no: 1603730758, highest available: 1603730757
suit: validated sequence number
)Formatted component name:
Comparing manifest offset 1000 with other slot offset
Comparing manifest offset 40800 with other slot offset
suit_parse() failed. res=-4

```

**Comment by fjmolinas** [comment]
> `examples/suit_update` is not passing for me on `iotlab-m3`, it fails with the following error, any ideas?
> 
> ```
> make[1]: Leaving directory '/home/francisco/workspace/RIOT3/examples/suit_update'
> suit: received URL: "coap://[fd00:dead:beef::1]/fw/iotlab-m3/suit_update-riot.suit_signed.1603730758.bin"
> suit_coap: trigger received
> suit_coap: downloading "coap://[fd00:dead:beef::1]/fw/iotlab-m3/suit_update-riot.suit_signed.1603730758.bin"
> suit_coap: got manifest with size 503
> suit: verifying manifest signature
> suit: validated manifest version
> )riotboot_hdr_validate: riotboot_hdr magic number invalid
> Manifest seq_no: 1603730758, highest available: 1603730757
> suit: validated sequence number
> )Formatted component name:
> Comparing manifest offset 1000 with other slot offset
> Comparing manifest offset 40800 with other slot offset
> suit_parse() failed. res=-4
> ```

I don't get the same behaviour on `samr21-xpro` though..

**Comment by bergzand** [comment]
> I don't get the same behaviour on `samr21-xpro` though..

It also works fine for me on the nrf52840dk

**Pull request merged**